### PR TITLE
Sprint15 cauchy laplace dists

### LINKED ISSUE #11: Add Laplace and Cauchy to CUQIpy-PyTorch

### PR DESCRIPTION
**Pull request by nabriis**
Closes #11 #12 

Also adds a demo showing how to use these distributions e.g. for non Gaussian likelihoods.

### CODE REVIEW

**Comment by nabriis** [comment]
Hi @amal-ghamdi, could you take a look at this PR? It adds two new wrappers for distributions in the pytorch framework and fixes a bug.

File: cuqipy_pytorch/distribution.py [hunk 1]
```diff
@@ -395,3 +395,109 @@ def gradient(self, v1, v2=None):
...
+
+            # Define scale value
+            scale = self._scale
+            if isinstance(scale, numbers.Number):
+                scale = scale*torch.ones(self.dim)
+            if isinstance(scale, np.ndarray):
+                scale = torch.tensor(scale)
+
+            # If both are tensors we create dist
+            if torch.is_tensor(loc) and torch.is_tensor(scale):
```
Comment (amal-ghamdi) [inline cuqipy_pytorch/distribution.py:443]: What should happen in the else clause? should we add an error?

File: cuqipy_pytorch/distribution.py [hunk 2]
```diff
@@ -395,3 +395,109 @@ def gradient(self, v1, v2=None):
             Q = self(v2).logpdf(v1) # Forward pass
             Q.backward()            # Backward pass
             return v2.grad
+
+class Laplace(Distribution):
```
Comment (amal-ghamdi) [inline cuqipy_pytorch/distribution.py:399]: Can Add a docstring here to provide needed information for the user

File: cuqipy_pytorch/distribution.py [hunk 3]
```diff
@@ -395,3 +395,109 @@ def gradient(self, v1, v2=None):
...
+            if torch.is_tensor(loc) and torch.is_tensor(scale):
+                self._dist = torch.distributions.Laplace(loc, scale)
+
+    def logpdf(self, value):
+        return torch.sum(self._dist.log_prob(value))
+
+    def _sample(self, n):
+        return self._dist.sample(torch.Size((n,)))
+
+class Cauchy(Distribution):
```
Comment (amal-ghamdi) [inline cuqipy_pytorch/distribution.py:452]: Here as well, can add a docstring

File: cuqipy_pytorch/model.py [hunk 4]
```diff
@@ -26,9 +26,12 @@ def backward(ctx, grad_output):
...
             return input.new(grad_input)
 
-    torch_forward = CustomFunction.apply
+    torch_forward = lambda *args: CustomFunction.apply(*args)
 
     # Add to cuqi model
     torch_model = Model(torch_forward, int(model.range_dim), int(model.domain_dim))
 
+    # Set parameter name
+    torch_model._non_default_args = model._non_default_args
```
Comment (amal-ghamdi) [inline cuqipy_pytorch/model.py:35]: seems a compact fix 👍

File: demos/demo_non_Gauss_Likelihood.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
+# %%
```
Comment (amal-ghamdi) [inline demos/demo_non_Gauss_Likelihood.py:1]: May be can add a brief description like:
"Demonstration showing a gradient based sampling using NUTs for two different Bayesian problems, both with non-Gaussian likelihoods."

File: demos/demo_non_Gauss_Likelihood.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+print(P1.logpdf(x=xp.ones(n)))
+print(P2.logpdf(u=xp.ones(n)))
+
+# %% Sampling
+
+# A convenience function to sample a Bayesian model
+# It creates the joint and posterior distributions as above.
+def sample_posterior(*densities, n_samples=500, n_burnin=500, **data):
+    """ Sample posterior given by a list of densities. The observations are given as keyword arguments. """
+    P = cuqi.distribution.JointDistribution(*densities)
```
Comment (amal-ghamdi) [inline demos/demo_non_Gauss_Likelihood.py:51]: It seems to me that creating the Joint distribution is happening twice in this demo. Can we avoid this or explain briefly why we do it twice, if for illustration purpose. (e.g. line 32 and line 51)

File: cuqipy_pytorch/distribution.py [hunk 7]
```diff
@@ -395,3 +395,109 @@ def gradient(self, v1, v2=None):
...
+
+            # If both are tensors we create dist
+            if torch.is_tensor(loc) and torch.is_tensor(scale):
+                self._dist = torch.distributions.Cauchy(loc, scale)
+
+    def logpdf(self, value):
+        return torch.sum(self._dist.log_prob(value))
+
+    def _sample(self, n):
+        return self._dist.sample(torch.Size((n,)))
```
Comment (amal-ghamdi) [inline cuqipy_pytorch/distribution.py:503]: May be an issue can be added to address code redundancy in some distribution wrappers to be considered in the future.

**Review by amal-ghamdi — CHANGES_REQUESTED** [review]
Thank you @nabriis for this PR. It looks great! I only have minor comments here and there.

File: cuqipy_pytorch/distribution.py (hunk 1 above)
Comment (nabriis) [reply cuqipy_pytorch/distribution.py:461]: We should not. Nothing should happen e.g. if its a lambda function

File: cuqipy_pytorch/distribution.py (hunk 7 above)
Comment (nabriis) [reply cuqipy_pytorch/distribution.py:538]: #14

**Comment by nabriis** [comment]
Hi @amal-ghamdi . Thanks for the review. I took the comments into account

File: cuqipy_pytorch/distribution.py (hunk 7 above)
Comment (amal-ghamdi) [reply cuqipy_pytorch/distribution.py:538]: Great!

**Review by amal-ghamdi — APPROVED** [review]
Thank you @nabriis. LGTM!
